File: solution/1971. Find if Path Exists in Graph/1971_02.cpp
class Solution {
public:
    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
        father.resize(n);
        for (int i = 0; i < n; ++i)
            father[i] = i;
        
        for (auto& e : edges)
            merge(e[0], e[1]);
        
        return findFather(source) == findFather(destination);
    }
    
private:
    vector<int> father;
    
    int findFather(int x) {
        if (x != father[x])
            father[x] = findFather(father[x]);
        return father[x];
    }
    
    void merge(int a, int b) {
        a = findFather(a), b = findFather(b);
        if (a < b)
            father[b] = a;
        else
            father[a] = b;
    }
};

File: solution/1971. Find if Path Exists in Graph/1971_02.py
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        father = [0] * n
        for i in range(n):
            father[i] = i
        
        def find_father(x):
            if x != father[x]:
                father[x] = find_father(father[x])
            return father[x]
        
        def merge(a, b):
            a, b = find_father(a), find_father(b)
            if a < b:
                father[b] = a
            else:
                father[a] = b
        
        for e in edges:
            merge(e[0], e[1])
            
        return find_father(source) == find_father(destination)
